src/coreComponents/physicsSolvers/multiphysics/SinglePhaseReservoirAndWells.cpp

```diff
@@ -33,46 +32,102 @@ namespace geosx
 using namespace dataRepository;
 using namespace constitutive;
 
-SinglePhaseReservoir::SinglePhaseReservoir( const string & name,
-                                            Group * const parent ):
-  ReservoirSolverBase( name, parent )
+namespace
 {
-  m_linearSolverParameters.get().mgr.strategy = LinearSolverParameters::MGR::StrategyType::singlePhaseReservoirFVM;
+
+template< typename SINGLEPHASE_RESERVOIR_SOLVER > class
+  SinglePhaseCatalogNames {};
+
+template<> class SinglePhaseCatalogNames< SinglePhaseFVM< SinglePhaseBase > >
+{
+public:
+  static string name() { return "SinglePhaseReservoirFVM"; }
+};
+template<> class SinglePhaseCatalogNames< SinglePhaseHybridFVM >
+{
+public:
+  static string name() { return "SinglePhaseReservoirHybridFVM"; }
+};
+
 }
 
-SinglePhaseReservoir::~SinglePhaseReservoir()
+// provide a definition for catalogName()
+template< typename SINGLEPHASE_RESERVOIR_SOLVER >
+string
+SinglePhaseReservoirAndWells< SINGLEPHASE_RESERVOIR_SOLVER >::
+catalogName()
+{
+  return SinglePhaseCatalogNames< SINGLEPHASE_RESERVOIR_SOLVER >::name();
+}
+
+template< typename SINGLEPHASE_RESERVOIR_SOLVER >
+SinglePhaseReservoirAndWells< SINGLEPHASE_RESERVOIR_SOLVER >::
+SinglePhaseReservoirAndWells( const string & name,
+                              Group * const parent )
+  : Base( name, parent )
+{}
+
+template< typename SINGLEPHASE_RESERVOIR_SOLVER >
+SinglePhaseReservoirAndWells< SINGLEPHASE_RESERVOIR_SOLVER >::
+~SinglePhaseReservoirAndWells()
 {}
 
-void SinglePhaseReservoir::initializePostInitialConditionsPreSubGroups()
+template< typename SINGLEPHASE_RESERVOIR_SOLVER >
+void
+SinglePhaseReservoirAndWells< SINGLEPHASE_RESERVOIR_SOLVER >::
+postProcessInput()
 {
-  ReservoirSolverBase::initializePostInitialConditionsPreSubGroups();
+  Base::postProcessInput();
 
-  if( m_flowSolver->getLinearSolverParameters().mgr.strategy == LinearSolverParameters::MGR::StrategyType::singlePhaseHybridFVM )
+  SINGLEPHASE_RESERVOIR_SOLVER * reservoirSolver =
+    &this->getParent().template getGroup< SINGLEPHASE_RESERVOIR_SOLVER >( m_reservoirSolverName );
+  std::get< toUnderlying< typename Base::SolverType >( Base::SolverType::Reservoir ) >( m_solvers ) = reservoirSolver;
+
+  SinglePhaseWell * wellSolver =
+    &this->getParent().template getGroup< SinglePhaseWell >( m_wellSolverName );
+  std::get< toUnderlying< typename Base::SolverType >( Base::SolverType::Well ) >( m_solvers ) = wellSolver;
```

Comment:
Thanks, I implemented the first part with the `setSubSolvers` functions in the base `CoupledSolver` class. For now it is simply used as 
https://github.com/GEOSX/GEOSX/blob/1dc1cf9ca37833654eefc891eaad228b6dd03dc1/src/coreComponents/physicsSolvers/multiphysics/CoupledReservoirAndWellsBase.cpp#L80

Maybe the first of to encapsulate the handling of the sub-solver names in the `CoupledSolver` is a little too error prone? I like the second one, I could do something like 
```
class WellSolverBase
{
  static string coupledSolverAttributePrefix() { return "well"; }
};
```
and then both the flow solvers and the poromechanics solvers would have:
```
static string coupledSolverAttributePrefix() { return "reservoir"; }
```
because my intent is to be able to use either pure flow or coupled poromechanics solvers as `RESERVOIR_SOLVER` here.

Reply:
Done, I implemented the second part of your suggestion as well, with the `SolverType::coupledSolverAttributePrefix()`. For now, I kept `static string coupledSolverAttributePrefix() { return "flow"; }` in the flow solvers to avoid a change in the xml.